wandb/sdk/service/streams.py

```diff
@@ -247,41 +242,51 @@ def _process_drop(self, action: StreamAction) -> None:
                 stream.drop()
                 stream.join()
 
-    def _on_probe_exit(self, probe_handle: MailboxProbe, stream: StreamRecord) -> None:
-        handle = probe_handle.get_mailbox_handle()
-        if handle:
-            result = handle.wait(timeout=0, release=False)
-            if not result:
-                return
-            probe_handle.set_probe_result(result)
-        handle = stream.interface.deliver_poll_exit()
-        probe_handle.set_mailbox_handle(handle)
-
-    def _on_progress_exit(self, progress_handle: MailboxProgress) -> None:
-        pass
-
-    def _on_progress_exit_all(
+    async def _finish_all_progress(
         self,
         progress_printer: progress.ProgressPrinter,
-        progress_all_handle: MailboxProgressAll,
+        streams_to_watch: dict[str, StreamRecord],
     ) -> None:
-        probe_handles: list[MailboxProbe] = []
-        progress_handles = progress_all_handle.get_progress_handles()
-        for progress_handle in progress_handles:
-            probe_handles.extend(progress_handle.get_probe_handles())
-
-        assert probe_handles
-
-        if self._check_orphaned():
-            self._stopped.set()
-
-        poll_exit_responses: list[pb.PollExitResponse] = []
-        for probe_handle in probe_handles:
-            result = probe_handle.get_probe_result()
-            if result:
-                poll_exit_responses.append(result.response.poll_exit_response)
-
-        progress_printer.update(poll_exit_responses)
+        """Poll the streams and display statistics about them.
+
+        This never returns and must be cancelled.
+
+        Args:
+            progress_printer: Printer to use for displaying finish progress.
+            streams_to_watch: Streams to poll for finish progress.
+        """
+        results: dict[str, pb.Result | None] = {}
+
+        async def loop_poll_stream(
+            stream_id: str,
+            stream: StreamRecord,
+        ) -> NoReturn:
+            while True:
+                start_time = time.monotonic()
+
+                handle = stream.interface.deliver_poll_exit()
+                results[stream_id] = await handle.wait_async(timeout=None)
+
+                elapsed_time = time.monotonic() - start_time
+                if elapsed_time < 1:
+                    await asyncio.sleep(1 - elapsed_time)
```

Comment:
is this current solution to allow to poll every second?

Reply:
This is to avoid polling too frequently. The old mailbox implements polling itself and polls about once a second. What API do you envision?

Reply:
oh was just thinking something that has the main function that you pass and this sleep time, so basically a wrapper around the pattern that you already have in multiple places

Reply:
I think it's possible, but I'll need to think more about a good design here, and I didn't want to spend too much time considering that `streams.py` is in `legacy-service`.

Reply:
fair!